adding new reactor concept (#101)
Basis-Theory/basistheory-integration-guides: collections/concepts/what-are-reactors.md

</span>

## **How It Works**

Reactors are serverless function runtimes, similar to AWS Lambda, Azure Functions, or Cloudflare Workers - except your applications, systems, and infrastructure never touches the sensitive plaintext data. 

<img src="/assets/images/concepts/reactors_overview.png">

## Reactor Formulas

Using a Reactor starts with the code you’d like to execute within your system, this code template is referred to as a Reactor Formula. When creating a Reactor Formula you will define the `code`, expected request parameters, and any up-front configuration - like secrets - required to execute your function. Find *an exhaustive list of all configuration options in our [Reactor Formula API Reference.](https://docs.basistheory.com/?shell#reactor-formulas-reactor-formula-object)*

Formulas provide two different response types, giving the flexibility of securing data or returning it as a raw response:

- `tokenize`
    - Any object passed will be [tokenized](https://docs.basistheory.com/#tokenize).
- `raw`
    - Any object passed will be returned from the HTTP request that invoked the Reactor.

```js
module.exports = async function (req) {
  // code to execute
  return {

>     tokenize: {
>       bar: 'foo' // will be tokenized
>     },
>     raw: {
>       foo: 'bar' // will be returned in plaintext
>     }
>   };

Author: ```suggestion
    tokenize: {
      bar: 'secret' // will be tokenized
    },
    raw: {
      foo: 'plaintext' // will be returned in plaintext
    }
  };
```

};
```

<span class="base-alert info">
  <span>

>    **Quickstart hint:** Find a list of our fully compliant and maintained Official Formulas from within your [Web Portal](https://portal.basistheory.com/reactors/formulas)

Reviewer: ```suggestion
   **Quickstart hint:** Find a list of our fully compliant and maintained Official Reactor Formulas from within our [Portal](https://portal.basistheory.com/reactors/formulas)
```

```js
module.exports = async function (req) {
  const fetch = require('node-fetch');
  const { customer_id } = req.args;
  const last4 = customer_id.substring(-4);
  
  const response = await fetch('https://httpbin.org/post', {method: 'POST', body: last4});
  const raw = await resp.json()

  return {
    raw,
  };
};
```

### Create a pdf document

Creating documents out of sensitive data is a primary need for businesses today, especially in fintech where you need to create and submit 1099s for many businesses:

```js
module.exports = async function (req) {
  const fetch = require('node-fetch');
  const PDFDocument = require('pdfkit');
  const { token: { data } } = req.args;
  
  let doc = new PDFDocument;
  doc.fontSize(8)
    .text(`Some token data on a pdf: ${data}`, 1, 1);
  doc.end();
  
  //Send or upload file to your partner
  const response = await fetch('https://httpbin.org/post', {method: 'POST', body: doc});
  const raw = await resp.json()

  return {
    raw,
  };
};
```

### Generate a text file and send to an SFTP server

> Many legacy business process still rely heavily on comma delimited files (CVS), tab delimited files or space-delimited files to transport data between companies, typically using SFTP servers as the endpoint of this data. For example, engaging with partner banks with ACH files requires you to format your file correctly and drop it on to an SFTP server. 

Reviewer: ```suggestion
Many legacy business process still rely heavily on comma delimited files (CSV), tab delimited files or space-delimited files to transport data between companies, typically using SFTP servers as the endpoint of this data. For example, engaging with partner banks with ACH files requires you to format your file correctly and drop it on to an SFTP server. 
```

        
        writeStream.on('close',function () {
          res.json({success: true});
          resolve();
        });
        bufferStream.pipe(writeStream);
      });
    }).connect(connSettings));
  };
```

### Import file from a partner

When you need to process files of sensitive data without it touching your systems, use a Reactor to desensitize a file before forwarding it on to your systems for your own logic:

```js
module.exports = async function (req) {
  const { BasisTheory } = require('@basis-theory/basis-theory-js');
  const { fileString } = req.args; // "name,ssn\nTheory,555445555"
  const { BT_API_KEY } = req.configuration;

  const bt = await new BasisTheory().init(BT_API_KEY);
  const rows = fileString.split("\n").map(r => r.split(","));

  await Promise.all(rows.slice(1).map(row => {
    return bt.tokens.create({
      type: 'social_security_number',
      data: row[1],
    }).then(token => row[1] = token.id);
  }));

  const desensitizedFile = rows.map(row => row.join(",")).join("\n");

  res.json({ raw: desensitizedFile}); // "name,ssn\nTheory,14f5c8d4-3e84-4185-96ac-fac6386614dc"
};
```

### Anything you can imagine

> When our templates and examples aren’t enough, we enable you to build anything you want to with our Reactors. Start with a blank function like the below and solve any business problem with the data you need:

Reviewer: ```suggestion
When our templates and examples aren’t enough, we enable you to build anything you want to with our Reactors. Start with a blank function like the one below and solve any business problem with the data you need:
```

```js
module.exports = async function (req) {
  const { tokens } = req.args;

  // Anything you can dream up 💭

  return {
    tokenize: { foo: "bar" }, // tokenize data
    raw: { foo: "bar"} // return any data
  };
};
```

---

## FAQ:

### When do I use a Reactor?

When you’re required to write your own code to solve more complex problems - for example when needing to manipulate the data, create a document, or import a file from a partner.

### When would I use the Proxy instead of a Reactor?

When making a simple HTTP request, a simpler implementation can be created using the Proxy. 

### What does the development lifecycle look like for building Reactors?

> Each Reactor runs a single function which can be scoped, coded, and tested all within your normal development tooling and lifecycles. Code written and pushed to your own Github repositories can be used to create new Reactor Formulas using our Terraform or [API](https://docs.basistheory.com/#reactor-formulas-reactor-formula-object) integrations.

Reviewer: ```suggestion
Each Reactor runs a single function which can be scoped, coded, and tested all within your normal development tooling and lifecycles. Code written and pushed to your own Github repositories can be used to create new Reactor Formulas using our Terraform Provider or [API](https://docs.basistheory.com/#reactor-formulas-reactor-formula-object) integrations.
```
